vm-virtio/src/net_util.rs

```diff
@@ -0,0 +1,223 @@
+// Copyright (c) 2019 Intel Corporation. All rights reserved.
+//
+// SPDX-License-Identifier: Apache-2.0 AND BSD-3-Clause
+
+use super::{DescriptorChain, DeviceEventT, Queue};
+use net_util::{MacAddr, Tap, TapError, MAC_ADDR_LEN};
+use std::cmp;
+use std::io::Write;
+use std::mem;
+use std::net::Ipv4Addr;
+use virtio_bindings::bindings::virtio_net::*;
+use vm_memory::{Bytes, GuestAddress, GuestMemoryMmap};
+
+type Result<T> = std::result::Result<T, Error>;
+
+/// The maximum buffer size when segmentation offload is enabled. This
+/// includes the 12-byte virtio net header.
+/// http://docs.oasis-open.org/virtio/virtio/v1.0/virtio-v1.0.html#x1-1740003
+const MAX_BUFFER_SIZE: usize = 65562;
+
+// A frame is available for reading from the tap device to receive in the guest.
+pub const RX_TAP_EVENT: DeviceEventT = 2;
+// The guest has made a buffer available to receive a frame into.
+pub const RX_QUEUE_EVENT: DeviceEventT = 0;
+// The transmit queue has a frame that is ready to send from the guest.
+pub const TX_QUEUE_EVENT: DeviceEventT = 1;
+// The device has been dropped.
+pub const KILL_EVENT: DeviceEventT = 3;
+// The device should be paused.
+pub const PAUSE_EVENT: DeviceEventT = 4;
+// Number of DeviceEventT events supported by this implementation.
+pub const NET_EVENTS_COUNT: usize = 5;
+
+#[derive(Debug)]
+pub enum Error {
+    /// Open tap device failed.
+    TapOpen(TapError),
+    /// Setting tap IP failed.
+    TapSetIp(TapError),
+    /// Setting tap netmask failed.
+    TapSetNetmask(TapError),
+    /// Setting tap interface offload flags failed.
+    TapSetOffload(TapError),
+    /// Setting vnet header size failed.
+    TapSetVnetHdrSize(TapError),
+    /// Enabling tap interface failed.
+    TapEnable(TapError),
+}
+
+#[derive(Clone)]
+pub struct TxVirtio {
+    pub iovec: Vec<(GuestAddress, usize)>,
+    pub frame_buf: [u8; MAX_BUFFER_SIZE],
+}
+
+impl Default for TxVirtio {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl TxVirtio {
+    pub fn new() -> Self {
+        TxVirtio {
+            iovec: Vec::new(),
+            frame_buf: [0u8; MAX_BUFFER_SIZE],
+        }
+    }
+
+    pub fn process_desc_chain(&mut self, mem: &GuestMemoryMmap, tap: &mut Tap, queue: &mut Queue) {
+        while let Some(avail_desc) = queue.iter(&mem).next() {
+            let head_index = avail_desc.index;
+            let mut read_count = 0;
+            let mut next_desc = Some(avail_desc);
+
+            self.iovec.clear();
+            while let Some(desc) = next_desc {
+                if desc.is_write_only() {
+                    break;
```

Comment:
If this happens, because we don't want to loose/waste the descriptor, we need to put it back in the list, invoking `queue.go_to_previous_position()` before the `break` statement.

Reply:
if the desc is write only, it means the device is not allowed to read the data there. Why we need to put it back in the list?